app/Services/UserService.php

```diff
@@ -0,0 +1,381 @@
+     */
+    public function createUser(array $input): User
+    {
+        // Make sure the email is lowercased
+        if (isset($input['email'])) {
+            $input['email'] = strtolower($input['email']);
+        }
+
+        // Validate input
+        $validated = Validator::make($input, [
+            'name' => [
+                'required',
+                'string',
+                'min:3',
+                'max:255',
+            ],
+            'email' => [
+                'required',
+                'string',
+                'email:filter',
+                'max:255',
+                new Lowercase(),
+                'unique:users,email',
+            ],
+            'password' => [
+                'required',
+                'string',
+                Rules\Password::defaults(),
+            ],
+            'language' => [
+                'required',
+                'string',
+                new Rules\Enum(Language::class),
+            ],
+        ])->validate();
+
+        $user = User::create([
+            'name'                               => $validated['name'],
+            'email'                              => $validated['email'],
+            'password'                           => Hash::make($validated['password']),
+            'language'                           => $validated['language'],
+            'email_verified_at'                  => null,
+            'email_verification_code'            => rand(100000, 999999),
+            'email_verification_code_expires_at' => now()->addHour(),
+            'remember_token'                     => Str::random(10),
+        ]);
+
+        Mail::to($user)->queue(new EmailVerification($user));
+
+        return $user;
+    }
+
+    /**
+     * Update certain user fields.
+     *
+     * @param User                  $user
+     * @param array<string, string> $input
+     *
+     * @return User
+     *
+     * @throws \Illuminate\Validation\ValidationException
+     */
+    public function patchUser(User $user, array $input): User
+    {
+        // Validate input
+        $validated = Validator::make($input, [
+            'name' => [
+                'sometimes',
+                'required',
+                'string',
+                'min:3',
+                'max:255',
+            ],
+            'language' => [
+                'sometimes',
+                'required',
+                'string',
+                new Rules\Enum(Language::class),
+            ],
+        ])->validate();
+
+        $user->update($validated);
+
+        return $user;
+    }
+
+    /**
+     * Upsert a user device.
+     *
+     * @param User                  $user
+     * @param string                $uuid
+     * @param array<string, string> $input
+     *
+     * @return Device
+     *
+     * @throws \Illuminate\Validation\ValidationException
+     */
+    public function upsertUserDevice(User $user, string $uuid, array $input): Device
```

Comment:
Missed to validate the `$uuid`. Is it intentional?

Reply:
Yes, it is intentional. Even though this is called uuid, it's not a UUIDv4. It's just a unique id that the device generates and exposes to the [device api](https://capacitorjs.com/docs/apis/device) on the frontend. It's just a string and we'll use it as it is. The validation already happens on the method signature, that will accept strings only (pretty much everything right?)